### DISCUSSION TITLE
Const functions

### DISCUSSION THREAD

**Discussion by JPaja** [dotnet/csharplang#5570] [Language Ideas] [answered]
## Idea

Primary idea is to add way to mark [pure functions](https://en.wikipedia.org/wiki/Pure_function) in some way so those functions can be evaluated in compile time when used in const fields.

## Solution

Possible syntax could be to allow `const` keyword in function declaration
```csharp
const uint Area(uint a, uint b, uint c) => 2 * (a * b + a * c + b * c);
const uint a = 1;
const uint b = 2;
const uint c = 3;
const uint area = Area(a,b,c); //Would evaluate to 22 in compile time
```
or for strings
```csharp
const uint Hash(ReadOnlySpan<char> value)
{
     uint s1 = 1;
     uint s2 = 0;

     for (int n = 0; n < value.Length; n++) {
        s1 = (s1 + value[n]) % 65521;
        s2 = (s2 + s1) % 65521;
     }     
     return (s2 << 16) | s1;
}

const string value = "SomeText";
const uint hashedValue = Hash(value); //Would evaluate to 236127034 in compile time
```

## Constraints
My current idea to ensure purity function must follow rules:
* Variables, Args, Return value can be only struct. 
* Functions, Properties and Fields used must be declared as const
* For properties  im not sure if to fully disable them, or allow get only properties.
* If those conditions are not met compiler would show errors

## Runtime
System.Math functions and similar could be flagged as const.

## Nice to have
Automatic flag for functions and properties that meet putiry rules.

### Functions that take too long or have infinite loop. 
- Solution 1 would be to have time out when evaluating const functions and error out when compiling
- Solution 2 would be to not allow loops, which would make Hash example not valid

### Functions that are not pure (possibly malicious) artificially flagged as pure in third party libraries.
- Solution be to check purity on IL level. If function is pure it shouldn't be able to RCE

**Comment by HaloFour** [comment, marked as answer]
See: #2379 #4803 #1028 #776 #504

**Comment by erik-kallen** [comment]
Why do we need a keyword for this? The compiler should be able to figure it out by itself.

**Reply by JPaja** [reply]
Yes but i think you should also be able to explicitly mark your function as pure so compiler can warn you out if you break the purity.

**Reply by jnm2** [reply]
Yes, especially because fluctuating back and forth is affecting downstream consumers of a library that provides such functions. Ideally breaking changes only happen when there are explicit changes at the API declaration site.
